devito/core/autotuning.py

```diff
@@ -222,38 +196,68 @@ def finalize_time_bounds(stepper, at_args, args, mode):
         args[dim.max_name] = args[dim.max_name]
 
 
-def more_heuristic_attempts(blocksizes):
-    # Ramp up to higher block sizes
-    handle = OrderedDict([(i, options['at_blocksize'][-1]) for i in blocksizes[0]])
-    for i in range(3):
-        new_bs = OrderedDict([(k, v*2) for k, v in handle.items()])
-        blocksizes.insert(blocksizes.index(handle) + 1, new_bs)
-        handle = new_bs
-
-    handle = []
-    # Extended shuffling for the smaller block sizes
-    for bs in blocksizes[:4]:
-        for i in blocksizes:
-            handle.append(OrderedDict(list(bs.items())[:-1] + [list(i.items())[-1]]))
-    # Some more shuffling for all block sizes
-    for bs in list(blocksizes):
-        ncombs = len(bs)
-        for i in range(ncombs):
-            for j in combinations(bs, i+1):
-                item = [(k, bs[k]*2 if k in j else v) for k, v in bs.items()]
-                handle.append(OrderedDict(item))
-
-    unique = []
-    for i in blocksizes + handle:
-        if i not in unique:
-            unique.append(i)
+def generate_block_shapes(blockable, args, level):
+    # Max attemptable block shape
+    max_bs = tuple((d.step.name, d.max_step.subs(args)) for d in blockable)
 
-    return unique
+    # Attempted block shapes:
+    # 1) Defaults (basic mode)
+    ret = [tuple((d.step.name, v) for d in blockable) for v in options['blocksize']]
+    # 2) Always try the entire iteration space (degenerate block)
+    ret.append(max_bs)
+    # 3) More attempts if auto-tuning in aggressive mode
+    if level == 'aggressive':
+        # Ramp up to larger block shapes
+        handle = tuple((i, options['blocksize'][-1]) for i, _ in ret[0])
+        for i in range(3):
+            new_bs = tuple((b, v*2) for b, v in handle)
+            ret.insert(ret.index(handle) + 1, new_bs)
+            handle = new_bs
+
+        handle = []
+        # Extended shuffling for the smaller block shapes
+        for bs in ret[:4]:
+            for i in ret:
+                handle.append(bs[:-1] + (i[-1],))
+        # Some more shuffling for all block shapes
+        for bs in list(ret):
+            ncombs = len(bs)
+            for i in range(ncombs):
+                for j in combinations(dict(bs), i+1):
+                    handle.append(tuple((b, v*2 if b in j else v) for b, v in bs))
+        ret.extend(handle)
+
+    # Drop unnecessary attempts:
+    # 1) Block shapes exceeding the iteration space extent
+    ret = [i for i in ret if all(dict(i)[k] <= v for k, v in max_bs)]
+    # 2) Redundant block shapes
+    ret = filter_ordered(ret)
```

Comment:
Can I make another request here?  Exclude block shapes where it would lead to having extremely few blocks to run (< the CPU count / 4 or something like that).  These almost-single-threaded autotuning steps are what seem to take the vast majority of the autotune time.